### PR TITLE
Add context biasing for mobile

### PR DESCRIPTION
**Pull request by ductranminh**
I implement context biasing (hot words) for android/ios frontend by calling:
```
Reset(hotWords)
```

### CODE REVIEW

File: android/SherpaOnnx/app/src/main/java/com/k2fsa/sherpa/onnx/SherpaOnnx.kt [hunk 1]
```diff
@@ -85,14 +85,17 @@ class SherpaOnnx(
         acceptWaveform(ptr, samples, sampleRate)
 
     fun inputFinished() = inputFinished(ptr)
-    fun reset(recreate: Boolean = false) = reset(ptr, recreate = recreate)
+    fun reset(recreate: Boolean = false, hotWords: String = "") = reset(ptr, recreate, hotWords)
```
Comment (csukuangfj) [inline android/SherpaOnnx/app/src/main/java/com/k2fsa/sherpa/onnx/SherpaOnnx.kt]: Could you also update
https://github.com/k2-fsa/sherpa-onnx/blob/558f5e32630e056f776ea2c3f3c3b1c062748410/android/SherpaOnnx2Pass/app/src/main/java/com/k2fsa/sherpa/onnx/SherpaOnnx.kt#L123

Comment (csukuangfj) [inline android/SherpaOnnx/app/src/main/java/com/k2fsa/sherpa/onnx/SherpaOnnx.kt]:
```suggestion
    fun reset(recreate: Boolean = false, hotwords: String = "") = reset(ptr, recreate, hotwords)
```

File: sherpa-onnx/jni/jni.cc [hunk 2]
```diff
@@ -76,11 +76,22 @@ class SherpaOnnx {
...
       stream_ = recognizer_.CreateStream();
-    } else {
+      } else {
       recognizer_.Reset(stream_.get());
+      }
+    } else {
+      auto stream = recognizer_.CreateStream(keywords);
+      // Set new keywords failed, the stream_ will not be updated.
+      if (stream != nullptr) {
+        stream_ = std::move(stream);
```
Comment (csukuangfj) [inline sherpa-onnx/jni/jni.cc:93]: Could you also use `SHERPA_ONNX_LOGE()` to print some logs here saying it failed to set the hotwords?

File: sherpa-onnx/jni/jni.cc [hunk 3]
```diff
@@ -1509,9 +1520,12 @@ JNIEXPORT void JNICALL Java_com_k2fsa_sherpa_onnx_SherpaOnnxOffline_delete(
 
 SHERPA_ONNX_EXTERN_C
 JNIEXPORT void JNICALL Java_com_k2fsa_sherpa_onnx_SherpaOnnx_reset(
-    JNIEnv *env, jobject /*obj*/, jlong ptr, jboolean recreate) {
+    JNIEnv *env, jobject /*obj*/, jlong ptr, jboolean recreate, jstring keywords) {
   auto model = reinterpret_cast<sherpa_onnx::SherpaOnnx *>(ptr);
-  model->Reset(recreate);
+  const char *p_keywords = env->GetStringUTFChars(keywords, nullptr);
+  std::string keywords_str = p_keywords;
+  model->Reset(recreate, keywords_str);
```
Comment (csukuangfj) [inline sherpa-onnx/jni/jni.cc]:
```suggestion
  model->Reset(recreate, p_keywords);
```

You can pass `p_keywords` directly to the function.

File: swift-api-examples/SherpaOnnx.swift [hunk 4]
```diff
@@ -237,8 +237,18 @@ class SherpaOnnxRecognizer {
...
+  /// If hotWords is not empty, it will create a new decoding stream with
+  /// the given hotWords appended to the default hotWords.
+  func reset(hotWords: String? = nil) {
+    guard let words = hotWords, !words.isEmpty else {
+        Reset(recognizer, stream)
+        return
+    }
+    
+    words.withCString { cString in
+        stream = CreateOnlineStreamWithHotwords(recognizer, cString)
```
Comment (csukuangfj) [inline swift-api-examples/SherpaOnnx.swift]: Please free the previous `stream` to avoid memory leak.

**Review by csukuangfj — COMMENTED** [review]
Thanks! 

Left some minor comments.

File: sherpa-onnx/jni/jni.cc (hunk 3 above)
Comment (ductranminh) [reply sherpa-onnx/jni/jni.cc:1527]: But `p_keywords` is `const char *` while the reset method accept `std::string`

Comment (ductranminh) [reply sherpa-onnx/jni/jni.cc]: But `p_keywords` is `const char *` while the reset method accept `std::string` ? It's incorrect type

**Comment by ductranminh** [comment]
@csukuangfj I updated the fixes

File: sherpa-onnx/jni/jni.cc (hunk 3 above)
Comment (csukuangfj) [reply sherpa-onnx/jni/jni.cc]: > But `p_keywords` is `const char *` while the reset method accept `std::string` ? It's incorrect type

Does it cause compiler errors? If yes, please post the error logs.

**Comment by csukuangfj** [comment]
> @csukuangfj I updated the fixes

Thanks! Only one comment left now. (I just added one.)

File: swift-api-examples/SherpaOnnx.swift [hunk 5]
```diff
@@ -237,8 +237,23 @@ class SherpaOnnxRecognizer {
 
   /// Reset the recognizer, which clears the neural network model state
   /// and the state for decoding.
-  func reset() {
-    Reset(recognizer, stream)
+  /// If hotWords is an empty string, it just recreates the decoding stream
+  /// If hotWords is not empty, it will create a new decoding stream with
+  /// the given hotWords appended to the default hotWords.
+  func reset(hotWords: String? = nil) {
```
Comment (csukuangfj) [inline swift-api-examples/SherpaOnnx.swift]:
```suggestion
  func reset(hotwords: String? = nil) {
```

Comment (ductranminh) [reply swift-api-examples/SherpaOnnx.swift]: done

File: sherpa-onnx/jni/jni.cc (hunk 3 above)
Comment (ductranminh) [reply sherpa-onnx/jni/jni.cc]: you're right, it works normally

File: sherpa-onnx/jni/jni.cc [hunk 6]
```diff
@@ -76,11 +76,24 @@ class SherpaOnnx {
...
+  // If keywords is not empty, it will create a new decoding stream with
+  // the given keywords appended to the default keywords.
+  void Reset(bool recreate, const std::string &keywords = {}) {
+    if (keywords.empty()) {
+      if (recreate) {
       stream_ = recognizer_.CreateStream();
-    } else {
+      } else {
       recognizer_.Reset(stream_.get());
+      }
```
Comment (csukuangfj) [inline sherpa-onnx/jni/jni.cc]: Could you indent the statements?

You can run
```
cd /path/to/sherpa-onnx
./scripts/check_style_cpplint.sh 
./scripts/check_style_cpplint.sh 1
./scripts/check_style_cpplint.sh 2
```
to check your code style.

**Review by csukuangfj — COMMENTED** [review]
Thanks! Looks good to me except for the C++ code style.

File: sherpa-onnx/jni/jni.cc (hunk 6 above)
Comment (ductranminh) [reply sherpa-onnx/jni/jni.cc]: > Could you indent the statements?
> 
> You can run
> 
> ```
> cd /path/to/sherpa-onnx
> ./scripts/check_style_cpplint.sh 
> ./scripts/check_style_cpplint.sh 1
> ./scripts/check_style_cpplint.sh 2
> ```
> 
> to check your code style.

Done!
All code style checks passed

**Comment by csukuangfj** [comment]
Thank you again for your contribution!

**Pull request merged**